Source/Core/InputCommon/InputConfig.h

```diff
@@ -4,29 +4,40 @@
 
 #pragma once
 
-#include <map>
+#include <memory>
+#include <string>
+#include <utility>
 #include <vector>
 
-#include "Common/FileUtil.h"
-#include "Common/IniFile.h"
-#include "Common/Thread.h"
-
-#include "InputCommon/ControllerEmu.h"
-#include "InputCommon/ControllerInterface/ControllerInterface.h"
+class ControllerEmu;
 
 class InputConfig
 {
 public:
-	InputConfig(const char* const _ini_name, const char* const _gui_name,
-		const char* const _profile_name)
-		: ini_name(_ini_name), gui_name(_gui_name), profile_name(_profile_name) {}
+	InputConfig(const std::string& ini_name, const std::string& gui_name, const std::string& profile_name)
+		: m_ini_name(ini_name), m_gui_name(gui_name), m_profile_name(profile_name)
+	{
+	}
 
 	bool LoadConfig(bool isGC);
 	void SaveConfig();
 
-	std::vector<ControllerEmu*>  controllers;
+	template <typename T, typename... Args>
+	void CreateController(Args&&... args)
+	{
+		m_controllers.emplace_back(std::make_unique<T>(std::forward<Args>(args)...));
+	}
+
+	ControllerEmu* GetController(int index);
```

Comment:
What do you think about:
```
template <typename T> T* GetControllerAs(int index)
{
  return static_cast<T*>(GetController(index));
}
```
Maybe it's also possible to verify the type within the base class, but I have no clue about this kind of c++ magic :/